src/lib/pgp-key.cpp

```diff
@@ -1404,13 +1423,65 @@ pgp_key_t::is_protected() const
 bool
 pgp_key_t::valid() const
 {
-    return validated_ && valid_;
+    return validity_.validated && validity_.valid && !validity_.expired;
 }
 
 bool
 pgp_key_t::validated() const
 {
-    return validated_;
+    return validity_.validated;
+}
+
+uint32_t
+pgp_key_t::valid_till_common(bool expiry) const
+{
+    if (!validated()) {
+        return 0;
+    }
+    uint32_t till = expiration() ? creation() + expiration() : 0xffffffff;
+    if (valid()) {
+        return till;
+    }
+    if (revoked()) {
+        /* we should not believe to the compromised key at all */
+        if (revocation_.code == PGP_REVOCATION_COMPROMISED) {
+            return 0;
+        }
+        const pgp_subsig_t &revsig = get_sig(revocation_.sigid);
+        if (revsig.sig.creation() > creation()) {
+            /* pick less time from revocation time and expiration time */
+            return std::min(revsig.sig.creation(), till);
+        }
+        return 0;
+    }
+    /* if key is not marked as expired then it wasn't valid at all */
```

Comment:
Why non-expired key cannot be valid here?

Reply:
@antonsviridenko Because we checked for validity on the line 1442, and if key is not expired and is not revoked then it wasn't valid, i.e. has no valid self-signature/binding signature.